crates/boundless-market/src/request_builder/finalizer.rs

```diff
@@ -0,0 +1,121 @@
+
+#[non_exhaustive]
+#[derive(Debug, Clone, Builder)]
+pub struct FinalizerConfig {
+    /// If true, the request's expiration time will be checked against the system clock.
+    #[builder(default = true)]
+    pub check_expiration: bool,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, Default)]
+pub struct Finalizer {
+    pub config: FinalizerConfig,
+}
+
+impl From<FinalizerConfig> for Finalizer {
+    fn from(config: FinalizerConfig) -> Self {
+        Self { config }
+    }
+}
+
+impl Default for FinalizerConfig {
+    fn default() -> Self {
+        Self::builder().build().expect("implementation error in Default for FinalizerConfig")
+    }
+}
+
+impl FinalizerConfig {
+    pub fn builder() -> FinalizerConfigBuilder {
+        Default::default()
+    }
+}
+
+impl Layer<(Url, RequestInput, Requirements, Offer, RequestId)> for Finalizer {
+    type Output = ProofRequest;
+    type Error = anyhow::Error;
+
+    async fn process(
+        &self,
+        (program_url, input, requirements, offer, request_id): (
+            Url,
+            RequestInput,
+            Requirements,
+            Offer,
+            RequestId,
+        ),
+    ) -> Result<Self::Output, Self::Error> {
+        let request = ProofRequest {
+            requirements,
+            id: request_id.into(),
+            imageUrl: program_url.into(),
+            input,
+            offer,
+        };
+
+        request.validate().context("built request is invalid; check request parameters")?;
+        if self.config.check_expiration && request.is_expired() {
+            bail!(
+                "request expired at {}; current time is {}",
+                request.expires_at(),
+                now_timestamp()
+            );
+        }
+        if self.config.check_expiration && request.is_lock_expired() {
+            bail!(
+                "request lock expired at {}; current time is {}",
+                request.lock_expires_at(),
+                now_timestamp()
+            );
+        }
+        Ok(request)
+    }
+}
+
+impl Adapt<Finalizer> for RequestParams {
+    type Output = ProofRequest;
+    type Error = anyhow::Error;
+
+    async fn process_with(self, layer: &Finalizer) -> Result<Self::Output, Self::Error> {
+        tracing::trace!("Processing {self:?} with Finalizer");
+
+        // We create local variables to hold owned values
+        let program_url = self.require_program_url()?.clone();
+        let input = self.require_request_input()?.clone();
+        let requirements = self.require_requirements()?.clone();
+        let offer = self.require_offer()?.clone();
+        let request_id = self.require_request_id()?.clone();
+
+        // As an extra consistency check. verify the journal satisfies the required predicate.
+        if let Some(ref journal) = self.journal {
+            if !requirements.predicate.eval(journal) {
+                bail!("journal in request builder does not match requirements predicate; check request parameters");
```

Comment:
Lets print the type, provided, and the expected values. We had an issue where Hans was trying to use this and it was hard to debug